### PR TITLE
app-editors/vis: Bump version to 0.4, add testing, fix eselect-vi dep’

### PR DESCRIPTION
**Pull request by lanodan**
Everything should be in the title, if you need more detailed commits see lanodanOverlay, I should be able to separate them if needed. (in the hope that Github isn’t borken)

Also I’m considering asking for ~arm on dev-libs/libtermkey and dev-libs/tre:= so this package can have ~arm KEYWORD with repoman being happy.

### CODE REVIEW

File: app-editors/vis/vis-0.4.ebuild [hunk 1]
```diff
@@ -3,21 +3,36 @@
 
 EAPI=5
 
+# Package test version and name
+MY_PTV=0.2
+MY_PTN=${PN}-test
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: This is used in only one place, so inline it.

File: app-editors/vis/vis-0.4.ebuild [hunk 2]
```diff
@@ -3,21 +3,36 @@
...
 #TODO: >=dev-lang/lua-5.2 (needed for syntax highlighting and settings)
-DEPEND=" dev-libs/libtermkey
+DEPEND=" app-eselect/eselect-vi
+	dev-libs/libtermkey
 	ncurses? ( sys-libs/ncurses:0= )
 	tre? ( dev-libs/tre:= )"
 RDEPEND="${DEPEND}"
 
+src_prepare() {
+	use test && (
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: Don't do pseudo-one-liners like this. Use a proper `if` block. Also, you're introducing unnecessary subshell. Also, you're lacking error handling for all the lines below.

File: app-editors/vis/vis-0.4.ebuild [hunk 3]
```diff
@@ -3,21 +3,36 @@
 
 EAPI=5
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: Bump to EAPI 6.

File: app-editors/vis/vis-0.4.ebuild [hunk 4]
```diff
@@ -3,21 +3,36 @@
...
+	dev-libs/libtermkey
 	ncurses? ( sys-libs/ncurses:0= )
 	tre? ( dev-libs/tre:= )"
 RDEPEND="${DEPEND}"
 
+src_prepare() {
+	use test && (
+		rm -r "${S}/test"
+		mv "${WORKDIR}/${MY_PT}" "${S}/test"
+		which vim &>/dev/null || sed -i 's/.*vim.*//' "${S}/test/Makefile"
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: `which` is a non-standard program that requires explicit dependency. Use `type -P` bash builtin instead.

**Comment by lanodan** [comment]
> There are too many broken packages to determine whether the breakages were added by the pull request. If in doubt, please rebase.

By rebase you mean pulling upstream and pushing to my fork? (Uh, pull requests…)

File: app-editors/vis/vis-0.3.ebuild [hunk 5]
```diff
@@ -3,13 +3,17 @@
 
 EAPI=5
 
+MY_PTV=0.1
+MY_PT=${PN}-test-${MY_PTV}
```
Comment (mgorny) [inline app-editors/vis/vis-0.3.ebuild]: Is there any reason to use `${PN}` here? It reduces readability, and is longer than plain `vis`…

File: app-editors/vis/vis-0.3.ebuild [hunk 6]
```diff
@@ -18,6 +22,14 @@ DEPEND=" dev-libs/libtermkey
 	tre? ( dev-libs/tre:= )"
 RDEPEND="${DEPEND}"
 
+src_prepare() {
+	if use test; then
+		rm -r "${S}/test"
```
Comment (mgorny) [inline app-editors/vis/vis-0.3.ebuild]: As I said, all those lines need error handling (`||die`). Also why are you using an absolute path when you're already inside `${S}`?

File: app-editors/vis/vis-9999.ebuild [hunk 7]
```diff
@@ -19,6 +19,12 @@ DEPEND=" dev-libs/libtermkey
 	tre? ( dev-libs/tre:= )"
 RDEPEND="${DEPEND}"
 
+src_prepare() {
+	if use test; then
+		type -P vim &>/dev/null || sed -i 's/.*vim.*//' "${S}/test/Makefile"
```
Comment (mgorny) [inline app-editors/vis/vis-9999.ebuild]: Both of the points apply here too.

File: app-editors/vis/vis-0.3.ebuild [hunk 8]
```diff
@@ -17,7 +17,8 @@ IUSE="+ncurses selinux test tre"
 
 #Note: vis is reported to also work with NetBSD curses
 #TODO: >=dev-lang/lua-5.2 (needed for syntax highlighting and settings)
-DEPEND=" dev-libs/libtermkey
+DEPEND=" app-eselect/eselect-vi
```
Comment (mgorny) [inline app-editors/vis/vis-0.3.ebuild]: Is it really a build-time dependency? I dare say it's only needed at runtime and `pkg_post*` which is also covered by runtime deps.

File: app-editors/vis/Manifest [hunk 9]
```diff
@@ -1 +1,2 @@
 DIST vis-0.3.tar.gz 334216 SHA256 bed8ed32ef4f81bc4fd28e297ea8edc80027ac6923b9d3b5bd955d24899c7077 SHA512 e1ca009bd7569f22dacf7464791f9f4499695d15d6e38e1be491d39e7ebe0530de643be5c1e7647e9beb7969a17efab6545d3e719fe8cf311562e89d636bba62 WHIRLPOOL a0ddd69b88b4131d4b9c2ba5494e255f20cd99f91998e4ad4fcec3e0298e39d9319571201b8ae4c8b4d21da78b99d865d72bfb534024a89863d66b9968e562c8
+DIST vis-test-0.1.tar.gz 95274 SHA256 78c2c4416ef284e2362bb0e761ee9c4add106d122ac44b341c6bcf449f0b17a6 SHA512 433a67892f27eb01f3ea6224e1ded393cd949b4d60f7c14656fdd526c02d2b309f63d39887d8d7ba11cbad3e183ab3cd1a913ac40aa29eadb7fa700eb76424b6 WHIRLPOOL 45d257cdc625d249e45d8581f11c976c4957812a41274943af622521e6a599bc012809fa17651dcc278579b1f0ef2d40468c746080cf2f82b52854854073e1e8
```
Comment (mgorny) [inline app-editors/vis/Manifest]: Squash this into the original commit needing this change. Don't introduce interim broken commits.

File: app-editors/vis/vis-0.4.ebuild [hunk 10]
```diff
@@ -1,9 +1,9 @@
 # Copyright 1999-2017 Gentoo Foundation
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: Please split version bump into separate commits, one adding the new version and the following one removing the old one, so that it's easier to revert the removal in case there are issues.

File: app-editors/vis/vis-0.4.ebuild [hunk 11]
```diff
@@ -1,9 +1,9 @@
...
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=5
+EAPI=6
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: It's generally considered a good idea to test the ebuilds you're submitting because this one will certainly fail hard.

File: app-editors/vis/Manifest [hunk 12]
```diff
@@ -1,2 +1,2 @@
-DIST vis-0.3.tar.gz 334216 SHA256 bed8ed32ef4f81bc4fd28e297ea8edc80027ac6923b9d3b5bd955d24899c7077 SHA512 e1ca009bd7569f22dacf7464791f9f4499695d15d6e38e1be491d39e7ebe0530de643be5c1e7647e9beb7969a17efab6545d3e719fe8cf311562e89d636bba62 WHIRLPOOL a0ddd69b88b4131d4b9c2ba5494e255f20cd99f91998e4ad4fcec3e0298e39d9319571201b8ae4c8b4d21da78b99d865d72bfb534024a89863d66b9968e562c8
-DIST vis-test-0.1.tar.gz 95274 SHA256 78c2c4416ef284e2362bb0e761ee9c4add106d122ac44b341c6bcf449f0b17a6 SHA512 433a67892f27eb01f3ea6224e1ded393cd949b4d60f7c14656fdd526c02d2b309f63d39887d8d7ba11cbad3e183ab3cd1a913ac40aa29eadb7fa700eb76424b6 WHIRLPOOL 45d257cdc625d249e45d8581f11c976c4957812a41274943af622521e6a599bc012809fa17651dcc278579b1f0ef2d40468c746080cf2f82b52854854073e1e8
+DIST vis-0.4.tar.gz 381046 SHA256 f11ba41cfb86dd39475960abfd12469de4da0ccfdb941f1d7680d89d987694c5 SHA512 d8fcf667ecad7b32752d6c5dbd004544e1a9283775d54d93c24ce8f314b98154aed5c6014cc03223d8427c2b1e3fd0d4b348dfa12ce30236fddd93b34521ee5f WHIRLPOOL 6e098bdac4aec74999a8555d3385a4a5312d797e495a1a25c36daa1287f3ada5410ad28b527e36c1a6710d4ef579166def32d6bcd4e5fd50bbed936e268f28a4
```
Comment (mgorny) [inline app-editors/vis/Manifest]: Squash into appropriate commits.

**Comment by mgorny** [comment]
> By rebase you mean pulling upstream and pushing to my fork? (Uh, pull requests…)

`git pull --rebase origin/master` works for most of the setups. Replace `origin` with whatever remote you're using for upstream Gentoo.

**Comment by lanodan** [comment]
> Is there any reason to use ${PN} here? It reduces readability, and is longer than plain vis…

vis will probably change of name in the future(martanne/vis#338), so I don’t think hardcoding the name is a good idea.

**Comment by mgorny** [comment]
>> Is there any reason to use ${PN} here? It reduces readability, and is longer than plain vis…
>
> vis will probably change of name in the future(martanne/vis#338), so I don’t think hardcoding the name is a good idea.

Except that this does not affect existing Gentoo distfiles at all, so having the name changed indirectly there is actually bad.

File: app-editors/vis/vis-9999.ebuild [hunk 13]
```diff
@@ -19,6 +19,12 @@ DEPEND=" dev-libs/libtermkey
 	tre? ( dev-libs/tre:= )"
 RDEPEND="${DEPEND}"
 
+src_prepare() {
+	if use test; then
+		type -P vim &>/dev/null || sed -i 's/.*vim.*//' "${S}/test/Makefile" || die
```
Comment (mgorny) [inline app-editors/vis/vis-9999.ebuild]: Don't use double `||`, it's unreadable.

File: app-editors/vis/vis-0.3.ebuild [hunk 14]
```diff
@@ -1,7 +1,7 @@
 # Copyright 1999-2017 Gentoo Foundation
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=5
+EAPI=6
```
Comment (mgorny) [inline app-editors/vis/vis-0.3.ebuild]: Since you are removing this version, there is no point in updating EAPI in it. The live ebuild is enough.

File: app-editors/vis/vis-0.4.ebuild [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+SRC_URI="https://github.com/martanne/vis/archive/v${PV}.tar.gz -> ${P}.tar.gz
+	test? ( https://github.com/martanne/vis/archive/v${MY_PTV}.tar.gz -> ${MY_PT}.tar.gz )"
+LICENSE="ISC"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+ncurses selinux test tre"
+
+#Note: vis is reported to also work with NetBSD curses
+#TODO: >=dev-lang/lua-5.2 (needed for syntax highlighting and settings)
+DEPEND=" dev-libs/libtermkey
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: Extraneous space after `"`.

**Comment by lanodan** [comment]
Sorry for the ~10 day hiatus, I had problems coping with how github pull request system works and ~breaks the git workflow, I think I’ll manage it more easily now.
I tried few things (src_prepare multi-if blocks, bumping EAPI at the end) but I hate no idea if they are better practice.

File: app-editors/vis/Manifest (hunk 9 above)
Comment (mgorny) [inline app-editors/vis/Manifest]:
```
!!! Fetched file: vis-test-0.1.tar.gz VERIFY FAILED!
!!! Reason: Filesize does not match recorded size
!!! Got:      221342
!!! Expected: 95274

File: app-editors/vis/Manifest [hunk 16]
```diff
@@ -1,2 +1,4 @@
...
 DIST vis-test-0.1.tar.gz 95274 SHA256 78c2c4416ef284e2362bb0e761ee9c4add106d122ac44b341c6bcf449f0b17a6 SHA512 433a67892f27eb01f3ea6224e1ded393cd949b4d60f7c14656fdd526c02d2b309f63d39887d8d7ba11cbad3e183ab3cd1a913ac40aa29eadb7fa700eb76424b6 WHIRLPOOL 45d257cdc625d249e45d8581f11c976c4957812a41274943af622521e6a599bc012809fa17651dcc278579b1f0ef2d40468c746080cf2f82b52854854073e1e8
+DIST vis-test-0.2.tar.gz 97770 SHA256 26f3a82842e098f4004f5a8741ac501e209aa36704971b7e68d8104102c90788 SHA512 892effa08c80c75e78eeb993ba8a0ddc110cc6f654fa1bf970454c35c839e9467b0ab41578f49af1fba0514c61f887fa044b9d01ab58f4dd749582a9550a2c29 WHIRLPOOL 9b45ab2bf0b8c434a367d4877748f2b1ef88f598e2e94d6f05c51366283852d163b7809a7e5e89f4923eb24e3488ee2714d1164da240647dd8af91108655ad6f
```
Comment (mgorny) [inline app-editors/vis/Manifest]:
```
!!! Fetched file: vis-test-0.2.tar.gz VERIFY FAILED!
!!! Reason: Filesize does not match recorded size
!!! Got:      241415
!!! Expected: 97770

**Comment by lanodan** [comment]
I just verified again that the files I got for the Manifest were the good ones, and they are and I can reproduce.

```
$ cat vis.pub
untrusted comment: vis editor signify public key
RWRbDa94LCndLy4pUdO6h1PmS1ooHOGb7p84OfQIR7+hFlZwuAXUdQ5J
$ qfile -v `which signify`
app-crypt/signify-20 (/usr/bin/signify)
$ signify -V -p vis.pub -x vis-test-0.2.tar.gz.sig -m vis-test-0.2.tar.gz
Signature Verified
$ signify -V -p vis.pub -x vis-test-0.1.tar.gz.sig -m vis-test-0.1.tar.gz
Signature Verified
```

**Comment by lanodan** [comment]
So, it has been more than a month…
Any news, solutions, or whatever else?

**Comment by mgorny** [comment]
There can be no progress if the package clearly doesn't work. Checksums don't match the files in `SRC_URI`. If the files are wrong, then fix `SRC_URI` to point to the right ones.

**Comment by lanodan** [comment]
I updated the Manifests for ``BLAKE2B``. Also the URL/Manifest issue was fixed on '26 Oct' for stating everything.

**Comment by lanodan** [comment]
@mgorny Any updates?
Everything should be fine since ~2 months ago now.

**Comment by mgorny** [comment]
I'm sorry. I've just tested and it doesn't seem to work:

```
>>> Emerging (3 of 3) app-editors/vis-0.4::gentoo
 * vis-0.4.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                               [ ok ]
 * vis-test-0.2.tar.gz BLAKE2B SHA512 size ;-) ...                                                                                          [ ok ]
>>> Unpacking source...
>>> Unpacking vis-0.4.tar.gz to /tmp/portage/app-editors/vis-0.4/work
>>> Unpacking vis-test-0.2.tar.gz to /tmp/portage/app-editors/vis-0.4/work
>>> Source unpacked in /tmp/portage/app-editors/vis-0.4/work
>>> Preparing source in /tmp/portage/app-editors/vis-0.4/work/vis-0.4 ...
mv: cannot move '/tmp/portage/app-editors/vis-0.4/work/' to a subdirectory of itself, 'test'
 * ERROR: app-editors/vis-0.4::gentoo failed (prepare phase):
 *   (no error message)
 * 
 * Call stack:
 *     ebuild.sh, line 124:  Called src_prepare
 *   environment, line 363:  Called die
 * The specific snippet of code:
 *           mv "${WORKDIR}/${MY_PT}" test || die;
 * 
 * If you need support, post the output of `emerge --info '=app-editors/vis-0.4::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=app-editors/vis-0.4::gentoo'`.
 * The complete build log is located at '/var/log/portage/app-editors:vis-0.4:20180127-092232.log'.
 * For convenience, a symlink to the build log is located at '/tmp/portage/app-editors/vis-0.4/temp/build.log'.
 * The ebuild environment file is located at '/tmp/portage/app-editors/vis-0.4/temp/environment'.
 * Working directory: '/tmp/portage/app-editors/vis-0.4/work/vis-0.4'
 * S: '/tmp/portage/app-editors/vis-0.4/work/vis-0.4'

```

**Comment by mgorny** [comment]
Even after fixing this, I see new issues with the new version:

1. invalid docdir + unnecessary LICENSE (this is a regression):
   ```
   >>> /usr/share/doc/vis/
   >>> /usr/share/doc/vis/README.md.lz
   >>> /usr/share/doc/vis/LICENSE.lz
   ```
2. econf used for custom configure script:
   ```
   ./configure --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --libdir=/usr/lib64 --enable-curses --disable-selinux --disable-tre
   ./configure: unknown option --host=x86_64-pc-linux-gnu
   ```
3. pre-stripped files:
   ```
    * QA Notice: Pre-stripped files found:
    * /usr/bin/vis-digraph
    * /usr/bin/vis-menu
    * /usr/bin/vis
   ```

2+3 are not a regression, so I can accept the PR with them. However, 1 needs to be fixed.

**Comment by lanodan** [comment]
Oh, my bad, the bug in ``src_prepare`` was a port error from my overlay.
Also the additionnal issues should be fixed, with each fix in a separate commit.

File: app-editors/vis/vis-0.4.ebuild [hunk 17]
```diff
@@ -25,7 +25,7 @@ RDEPEND="${DEPEND}
 src_prepare() {
 	if use test; then
 		rm -r test || die
-		mv "${WORKDIR}/${MY_PT}" test || die
+		mv "${WORKDIR}/vis-test-${MY_PT}" test || die
```
Comment (mgorny) [inline app-editors/vis/vis-0.4.ebuild]: This is still wrong. Also please squash this into the first commit, as it is a breakage introduced in it and there's no point adding the breakage in the first place.

**Comment by mgorny** [comment]
Thanks a lot for your patience and sorry for all the trouble you had to go through. The ebuild looks good, tests good, installs good and vis edits the commit messages good ;-).

**Comment by lanodan** [comment]
Thanks for your patience too. And well don’t be sorry, theses were my mistakes and I’m sorry for pushing commits that weren’t tested in the building side.
